### PR TITLE
Rule 11.4 improvements

### PR DESCRIPTION
**Pull request by lcartey**
## Description

This pull request makes the following improvements to the query for Rule 11.4:
 * Consider `0` to be a null pointer constant (fixes https://github.com/github/codeql-coding-standards/issues/331).
 * Report the actual types and order in the error message.
 * If the result arises from a macro invocation that is not function-like, then report the macro itself. This helps reduce the overall number of results, and reduces the duplication by not reporting expansions of the same macro multiple times.
 * If the result arises from a macro invocation which is function-like we now report the macro in addition to the regular result, to aid with debugging.

## Change request type

 - [ ] Release or process automation (GitHub workflows, internal scripts)
 - [ ] Internal documentation
 - [ ] External documentation
 - [x] Query files (`.ql`, `.qll`, `.qls` or unit tests)
 - [ ] External scripts (analysis report or other code shipped as part of a release)

## Rules with added or modified queries

 - [ ] No rules added
 - [ ] Queries have been added for the following rules:
     - _rule number here_
 - [x] Queries have been modified for the following rules:
     - `RULE-11-4`

## Release change checklist

A change note ([development_handbook.md#change-notes](https://github.com/github/codeql-coding-standards/blob/main/docs/development_handbook.md#change-notes)) is required for any pull request which modifies:

 - The structure or layout of the release artifacts.
 - The evaluation performance (memory, execution time) of an existing query.
 - The results of an existing query in any circumstance.

If you are only adding new rule queries, a change note is not required.

_**Author:**_ Is a change note required? 
  - [x] Yes
  - [ ] No

🚨🚨🚨
_**Reviewer:**_ Confirm that format of *shared* queries (not the .qll file, the
.ql file that imports it) is valid by running them within VS Code. 
  - [x] Confirmed

_**Reviewer:**_ Confirm that either a change note is not required or the change note is required and has been added.
  - [x] Confirmed

## Query development review checklist

For PRs that add new queries or modify existing queries, the following checklist should be completed by both the author and reviewer:

### Author

 - [x] Have all the relevant rule package description files been checked in?
 - [x] Have you verified that the metadata properties of each new query is set appropriately?
 - [x] Do all the unit tests contain both "COMPLIANT" and "NON_COMPLIANT" cases?
 - [x] Are the alert messages properly formatted and consistent with the [style guide](https://github.com/github/codeql-coding-standards/blob/main/docs/development_handbook.md#query-style-guide)?
 - [x] Have you run the queries on OpenPilot and verified that the performance and results are acceptable?<br />_As a rule of thumb, predicates specific to the query should take no more than 1 minute, and for simple queries be under 10 seconds. If this is not the case, this should be highlighted and agreed in the code review process._
 - [x] Does the query have an appropriate level of in-query comments/documentation?
 - [x] Have you considered/identified possible edge cases?
 - [x] Does the query not reinvent features in the standard library?
 - [x] Can the query be simplified further (not golfed!)

### Reviewer 

 - [x] Have all the relevant rule package description files been checked in?
 - [x] Have you verified that the metadata properties of each new query is set appropriately?
 - [x] Do all the unit tests contain both "COMPLIANT" and "NON_COMPLIANT" cases?
 - [x] Are the alert messages properly formatted and consistent with the [style guide](https://github.com/github/codeql-coding-standards/blob/main/docs/development_handbook.md#query-style-guide)?
 - [x] Have you run the queries on OpenPilot and verified that the performance and results are acceptable?<br />_As a rule of thumb, predicates specific to the query should take no more than 1 minute, and for simple queries be under 10 seconds. If this is not the case, this should be highlighted and agreed in the code review process._
 - [x] Does the query have an appropriate level of in-query comments/documentation?
 - [x] Have you considered/identified possible edge cases?
 - [x] Does the query not reinvent features in the standard library?
 - [x] Can the query be simplified further (not golfed!)

### CODE REVIEW

File: c/common/src/codingstandards/c/Pointers.qll [hunk 1]
```diff
@@ -62,12 +62,7 @@ class ArrayPointerArithmeticExpr extends PointerArithmeticExpr, ArrayExpr {
 predicate isNullPointerConstant(Expr e) {
   e.findRootCause() instanceof NULLMacro
   or
-  exists(CStyleCast c |
-    not c.isImplicit() and
-    c.getExpr() = e and
-    e instanceof Zero and
-    c.getType() instanceof VoidPointerType
-  )
+  e instanceof Zero
```
Comment (lcartey) [inline c/common/src/codingstandards/c/Pointers.qll:65]: Note to self: this affects Rule 11.9, which is incorrect. `0` is a null pointer constant, but not permitted according to 11.9.

File: change_notes/2023-07-28-rule-11-4-improvements.md [hunk 2]
```diff
@@ -0,0 +1,4 @@
+ - `RULE-11-4`
```
Comment (knewbury01) [inline change_notes/2023-07-28-rule-11-4-improvements.md:1]:
```suggestion
- `RULE-11-4` - `ConversionBetweenPointerToObjectAndIntegerType.ql`:
   - Fixed issue #331 - consider `0` a null pointer constant.
```

File: change_notes/2023-07-28-rule-11-4-improvements.md [hunk 3]
```diff
@@ -0,0 +1,4 @@
...
+   - Reduce false positives by considering `0` a null pointer constant.
```
Comment (knewbury01) [inline change_notes/2023-07-28-rule-11-4-improvements.md:2]:
```suggestion
```

**Review by knewbury01 — CHANGES_REQUESTED** [review]
minor comments only, otherwise looks good! @lcartey 

needs a merge resolve, 
and I think you accidentally marked the reviewer's checkboxes , not the author's ones, I agree with the completion of those, but please also mark the author ones!

File: c/common/src/codingstandards/c/Pointers.qll (hunk 1 above)
Comment (lcartey) [reply c/common/src/codingstandards/c/Pointers.qll]: I fixed the incompatibility with Rule 11.9, which didn't change the behaviour of Rule 11.9 itself. However, I did notice and fix a small issue, for which I have included a change note entry.

**Comment by lcartey** [comment]
@knewbury01 this has been updated, and ready for re-review.

**Review by knewbury01 — APPROVED** [review]
lgtm!

**Pull request merged**